Block spacing: block-level axial gap block support

ramonjd opened the pull request:
Resolves https://github.com/WordPress/gutenberg/issues/34529

## Description

I'm not 100% sure whether this is the right approach. For individual block supports it seems to do the job. Just throwing it out there for now.


## How has this been tested?

Using TwentyTwentyTwo, or another block based theme with `blockGap` support enabled in `theme.json`. 

In social-links/block.json, switch the value of `"supports.spacing.blockGap"` from `true` to `[ "vertical", "horizontal" ]`


**With axial**

https://user-images.githubusercontent.com/6458278/148323374-6d006c9e-c1de-45bf-b647-433d669499a1.mp4

**Without**

https://user-images.githubusercontent.com/6458278/148323368-d1babeab-19b8-4575-ba4f-11e5c9b50675.mp4

You can test with any appropriate block that supports `blockGap`, for example Navigation. Make sure to enable the support in the block.json file as described above.

<details><summary>Example editor code with blockGap as string value</summary>


  ```html
<!-- wp:heading {"level":3} -->
<h3 id="column">Column</h3>
<!-- /wp:heading -->

<!-- wp:columns {"style":{"spacing":{"blockGap":"167px"}}} -->
<div class="wp-block-columns"><!-- wp:column {"style":{"color":{"background":"#fee5d3"}}} -->
<div class="wp-block-column has-background" style="background-color:#fee5d3"><!-- wp:paragraph -->
<p>one</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column {"style":{"color":{"background":"#fee5d3"}}} -->
<div class="wp-block-column has-background" style="background-color:#fee5d3"><!-- wp:paragraph -->
<p>two</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column {"style":{"color":{"background":"#fee5d3"}}} -->
<div class="wp-block-column has-background" style="background-color:#fee5d3"><!-- wp:paragraph -->
<p>three</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column {"style":{"color":{"background":"#fee5d3"}}} -->
<div class="wp-block-column has-background" style="background-color:#fee5d3"><!-- wp:paragraph -->
<p>four</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column {"style":{"color":{"background":"#fee5d3"}}} -->
<div class="wp-block-column has-background" style="background-color:#fee5d3"><!-- wp:paragraph -->
<p>five</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:heading -->
<h2 id="group">Group</h2>
<!-- /wp:heading -->

<!-- wp:group {"style":{"color":{"background":"#fcdddd"}}} -->
<div class="wp-block-group has-background" style="background-color:#fcdddd"><!-- wp:paragraph -->
<p>I'm inside a group!</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>I'm inside a group!</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>I'm inside a group!</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:heading {"level":3} -->
<h3 id="navigation">Navigation</h3>
<!-- /wp:heading -->

<!-- wp:navigation {"ref":129,"style":{"spacing":{"blockGap":"22"}}} /-->

<!-- wp:heading {"level":3} -->
<h3 id="social-icons">Social icons</h3>
<!-- /wp:heading -->

<!-- wp:social-links {"style":{"spacing":{"blockGap":"4em"}}} -->
<ul class="wp-block-social-links"><!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->

<!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->

<!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->

<!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->

<!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /--></ul>
<!-- /wp:social-links -->
    
  ```


</details>

The above code uses strings values for `"blockGap"`. 

When toggling them they should covert to the style object `{"style":{"spacing":{"blockGap":{"top":"1.13em","left":"1.13em"}}}}`.

Point being, we should handle both string and object values seamlessly for backwards compatibility.


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
Initial commit: splitting the gap value to allow for axial gap values

ramonjd commented:
Just jotting down notes.

I have an open question about what to do with block-level theme.json properties.

The gap value is currently available only at the [top level](https://github.com/WordPress/gutenberg/blob/3f43af06db0b017bc09124f1ca99ed89d16cb471/lib/compat/wordpress-5.9/class-wp-theme-json-gutenberg.php#L431) and [assigned to a CSS VAR in PROPERTIES_METADATA](https://github.com/WordPress/gutenberg/blob/3f43af06db0b017bc09124f1ca99ed89d16cb471/lib/compat/wordpress-5.9/class-wp-theme-json-gutenberg.php#L182).

We're turning on blockGap support in individual blocks – see [Columns Block](https://github.com/WordPress/gutenberg/pull/37436) for example –  so I would imagine theme authors might want a way to control things at the block level in `theme.json` as well.

Put another way:

Should this toggle spacing block supports for an individual block?
```json
"settings": {
    "blocks": {
        "core/social-links": {
            "spacing": {
                "blockGap": true
            }
        }
    }
},

```
Should this determine the default value of block spacing for an individual block?
```json
"styles": {
    "blocks": {
        "core/social-links": {
            "spacing": {
                "blockGap": "20px"
            }
        }
    }
}

```

Also, even if we do allow global styles for blockgap, they're being overwritten by the CSS generated by layout.php support

<img width="427" alt="Screen Shot 2022-01-07 at 11 51 03 am" src="https://user-images.githubusercontent.com/6458278/148473473-45eab7a6-382d-4bd8-92fa-e0dae6c734c1.png">

ramonjd committed:
Checking for object. Now passing complete object to box control.

ramonjd committed:
Providing a default value for split gap values, where one is `undefined`.

andrewserong commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -128,17 +140,59 @@ export function GapEdit( props ) {
 		}
 	};
 
+	const blockGapValue = style?.spacing?.blockGap;
```
This function is getting pretty long now. Would it be worth splitting out the `boxValues` logic to a `getBoxValues` function? We could then maybe add a unit test for it, too, if it seems worthwhile?

ramonjd replied:
Done 👍

andrewserong commented on packages/block-library/src/columns/block.json:
```diff
@@ -28,7 +28,7 @@
 		},
 		"spacing": {
-			"blockGap": true,
+			"blockGap": [ "vertical", "horizontal" ],
```
Just a note that I think this is a deceptively powerful feature (by which I mean, very cool feature!) — because vertical spacing only affects the columns block on mobile viewports, this means that it creates a nice ability to be able to control desktop versus mobile spacing, without having to consider viewport widths directly 👍

youknowriad replied:
Which formats do we support here and what do they mean?

ramonjd replied:
> Which formats do we support here and what do they mean?

I'm not clear what you mean here. Was it in relation to the [above comment](https://github.com/WordPress/gutenberg/pull/37736#discussion_r804376196), or the formats of `"blockGap"` in `theme.json` (`string|array`)?

> because vertical spacing only affects the columns block on mobile viewports, this means that it creates a nice ability to be able to control desktop versus mobile spacing, without having to consider viewport widths directly

This comment makes me think of a recent conversation on how blockGap effects margin-top via the layout abstraction. Namely, how can we reconcile the need for zero margins in a world where `"blockGap"` has a value in theme.json.

https://github.com/WordPress/gutenberg/pull/37344

andrewserong reviewed (commented):
This is testing really well for me, @ramonjd, thanks so much for persevering with axial support!

Just left an optional comment about possibly moving some of the logic to a separate function to tighten thing up a little bit.

Overall I think this is a really clever approach — opting in to axial support for those blocks that use flex layout, where the gap value is used in the shorthand syntax, so concatenating the separate values together works nicely. This sounds like the kind of thing that we'd want to support long-term to me, but it might be worth confidence checking with other folks, too, that we're happy to rely on a (potential) implementation detail of the layout support? CC: @youknowriad in case there are any concerns there.

I really like the idea of starting out with this only supporting axial controls at the individual block level, and the global styles UI controls appear to be unaffected.

Tested with the social links, navigation, and columns blocks on a few different browsers, and using the TT2 theme.

The only really small issue I noticed (and I think this is to do with the `BoxControl` component's axial control support, rather than this PR) is that it doesn't say `mixed` when there's differing values (unlike, say, padding on the group block), but I don't think that's a blocker:

![Kapture 2022-02-11 at 17 02 31](https://user-images.githubusercontent.com/14988353/153544564-b1147514-5fb7-4026-b1d4-692f4cf3ff55.gif)

Otherwise, this is looking really good to me!

youknowriad commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -95,18 +97,28 @@ export function GapEdit( props ) {
+		if ( typeof next === 'object' ) {
+			const row = next?.top || 0;
+			const column = next?.left || 0;
+			newValue = row === column ? row : `${ row } ${ column }`;
```
Why are we storing the value as a concatenated string, shouldn't we store the next object as is like we do in other kind of properties (padding, margin...)? Is it because we only allow two sides? if that's the case should we instead use a `{ vertical, horizontal }` instead if `{ top, left }` is not good enough?

youknowriad commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -95,18 +97,28 @@ export function GapEdit( props ) {
+		if ( typeof next === 'object' ) {
+			const row = next?.top || 0;
+			const column = next?.left || 0;
+			newValue = row === column ? row : `${ row } ${ column }`;
```
Why are we storing the value as a concatenated string, shouldn't we store the next object as is like we do in other kind of properties (padding, margin...)? Is it because we only allow two sides? if that's the case should we instead use a `{ vertical, horizontal }` instead if `{ top, left }` is not good enough?

ramonjd replied:
It was out of convenience to use the CSS shorthand; mainly so I didn't have to alter the [layout implemention](https://github.com/WordPress/gutenberg/blob/a010546dbb95f55ba74b0d78c27b5a8f4e394167/lib/block-supports/layout.php#L95).

There might have been other reasons, but I can't recall them. I'll revisit and confirm. 👍 

I guess it might be better in the long run if the values were separate in case we want to use the values independently. Though ultimately the approach wasn't right, it's where we were trying to go with https://github.com/WordPress/gutenberg/pull/35454, which added separate CSS vars for row and column.

We could do a check for `is_array( $gap_value )` and, in that case, declare separate `row-gap` and `column-gap` values here. 

Definitely worth another look, thanks for nudge.

> if that's the case should we instead use a { vertical, horizontal } instead if { top, left } is not good enough?

`{ top, left }` should be good enough for our purposes for now. 

The `top, right, bottom, left` prop names are baked into the [Axial Input Controls](https://github.com/WordPress/gutenberg/blob/034cd9f299fbdf6afb3dcc62ad7986df3ff5a840/packages/components/src/box-control/axial-input-controls.js) in the [BoxControl component](https://github.com/WordPress/gutenberg/blob/034cd9f299fbdf6afb3dcc62ad7986df3ff5a840/packages/components/src/box-control/index.js) anyway.

ramonjd replied:
> Definitely worth another look, thanks for nudge.

Actually, definitely worth doing now that I've looked a little more closely.  We need to support `"blockGap": [ "vertical", "horizontal" ],` and `"blockGap": [ "vertical" ],` for example. 

There is no **"doh!"** emoji 😆

ramonjd replied:
> Why are we storing the value as a concatenated string, shouldn't we store the next object as is like we do in other kind of properties (padding, margin...)? 

523d9beff8a95e1d860cda289c33b14667342f54 is a very rough first stab at it.

I'm yet to address the question of the style model. It's made a smidge easier given that we're not adding `gap` inline CSS to block containers.

But if we have to match the CSS properties, I think we'd need to use the longhand properties.


```js
	setAttributes( {
		style: {
			...style,
			spacing: {
				...style?.spacing,
                                 rowGap: next?.top || 0,
                                 columnGap: next?.left || 0,
			},
		},
	} );

```

Which probably means changing a whole bunch of other references.

ramonjd committed:
This is the first commit that switches the blockGap over to use to an object. We're using the `top` and `left` properties of the BoxControl next value to represent row and column values.

Reverting some block.json changes.
Ensuring that we cater for blockGap string values.

Reverting some block.json changes.
Ensuring that we cater for blockGap string values.
Adding tests.

ramonjd commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -38,6 +39,45 @@ export function hasGapValue( props ) {
+	const column = blockGapBoxControlValue?.left || defaultValue;
+
+	return row === column ? row : `${ row } ${ column }`;
+}
```
Here I was wondering whether we should be returning the full definition so that we can isolate row and column values where only one is set, e.g., 

```js
// row and column is set or is the same
return `gap: 10px 10px;`;

// only row is set
return  `row-gap: 10px;`;

// only column is set
return  `column-gap: 10px;`;
```

That way, if a block ever has to inherit either row/column we're not overwriting it. 

Just typing out loud...

tellthemachines replied:
That would be a nice feature, though if we do it here we should do it on the server-side too 🙂

ramonjd replied:
After some further tinkering I probably need to put more thought into this.

I'm picturing a future where theme.json authors could add something like this.

```json
"spacing": {
    "blockGap": {
         "row": "0.5em",
         "column": "2em"
     }
}
```

and blocks would then inherit or use these values as defaults. 

Currently "blockGap" accept a string, which can be really anything, including the shorthand value of `"0.5em 2em"` so it might take some more cogitating to make things backwards compatible.

tellthemachines replied:
Sure, we don't need it for this PR; better to work out the details as a future enhancement.

ramonjd commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -38,6 +39,45 @@ export function hasGapValue( props ) {
+	return cleanEmptyObject( {
+		top: isValueString ? rawBlockGapValue : rawBlockGapValue?.top,
+		left: isValueString ? rawBlockGapValue : rawBlockGapValue?.left,
+	} );
```
The style representation on the block attribute will be

```json
{"style":{"spacing":{"blockGap":{"top":"1.13em","left":"1.13em"}}}}
```

I'm wondering if `top` and `left` make sense long term, where `row` and `column` might be more understandable.

I know `blockGap` is a special case, but now might be the time to decide whether we want the properties to represent the CSS spec or whether we just roll with what we've got.

andrewserong replied:
Looking at it again, I see what you mean — `row` and `column` have a better semantic meaning, too, given that `top` and `left` don't make much sense in the context of the `gap` CSS property 🤔

One of the things that I like about `vertical` and `horizontal` is that it's the name used for the shorthand form of margin and padding when only two values are used, but as a label, it too is slightly removed from being aligned to the CSS spec. I think I might be leaning toward `row` and `column`?

But I suppose, either way, if we use `row` or `vertical`, when we generate the layout styles, we'll know which one to grab to apply it to `margin-top` in the default layout, and to `row-gap` (or the shorthand syntax) in the flex layout 😃

ramonjd replied:
Yes, naming is hard 😆 

I think `row` and `column` are more intuitive in this context, but it's hard to judge whether it'll remain so in the future and, as you point out, in conjunction with other layout usages.

I tend to view the layout abstraction as in flux and something perhaps we shouldn't prioritize when it comes to naming block gap properties. What I mean to say is that layout should be smart enough to know about its own implementation, and it might change anyway :)

I too am leaning towards `row` and `column` only because the rightly describe the properties, and will correspond (sorta) to how the rest of the internet describes gap.

We're using `vertical` and `horizontal` for the axial descriptors in block.json only because that what the BoxControl is looking for. 

We could extend BoxControl to accept other axial-like values such as `row` and `column` if folks think it'd be more consistent. 

I'd hate to set something in stone now and have to roll back later.

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -66,6 +66,9 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
+			}
 			$gap_style = $gap_value ? $gap_value : 'var( --wp--style--block-gap )';
 			$style    .= "$selector > * { margin-top: 0; margin-bottom: 0; }";
 			$style    .= "$selector > * + * { margin-top: $gap_style;  margin-bottom: 0; }";
```
I know you're still exploring here, but just flagging that on this line `$gap_style` needs to be a single value, since it's passed in to `margin-top`.

ramonjd replied:
Thanks, totally missed that. 🙇

ramonjd committed:
margin-top should take a single value

ramonjd committed:
null coalescing operator (??) is not present in PHP

ramonjd committed:
Checking for gap array items before using them.

ramonjd committed:
Using `--wp--style--block-gap` as the fallback for split values

tellthemachines commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -38,6 +39,45 @@ export function hasGapValue( props ) {
+ */
+export function getGapValueFromStyle( rawBlockGapValue ) {
+	if ( ! rawBlockGapValue ) {
+		return rawBlockGapValue;
```
Hmm, any reason to return the value if it doesn't exist, as opposed to returning `null`?

ramonjd replied:
No reason at all 😄 I'll update and add a test. 👍

tellthemachines commented on packages/block-editor/src/hooks/gap.js:
```diff
@@ -128,17 +171,42 @@ export function GapEdit( props ) {
+		? {
+				...gapValue,
+				right: gapValue?.left,
+				bottom: gapValue?.top,
```
Might be nice to leave a comment here explaining why we're translating left to right and top to bottom 😅

ramonjd replied:
Good point, thanks. Will do.

tellthemachines reviewed (approved):
Code looks good and functionality is working well 🚢  Just some minor questions/comments below. Also:

> We're turning on blockGap support in individual blocks – see https://github.com/WordPress/gutenberg/pull/37436 for example – so I would imagine theme authors might want a way to control things at the block level in theme.json as well.

This would make sense. We don't necessarily want to be using the same gap for something small and inline like Social Icons as we would e.g. for Columns, so themes should be able to set this.

> Also, even if we do allow global styles for blockgap, they're being overwritten by the CSS generated by layout.php support

Could we somehow write the global style value to the `--wp--style--block-gap` variable? Or else perhaps the editor default should be at a lower level. 

In any case, that can be addressed as a separate PR 🙂

ramonjd committed:
Returning null instead of the argument in getGapCSSValue and updated tests
Added comment to explain how we build and pass the boxcontrol component value.

ramonjd committed:
Rolling back using the gap CSS var as a fallback since the CSS var can theoretically be something like "24px 54px" or some other manifestation of the shorthand value.

andrewserong reviewed (commented):
Thanks for persevering with this feature @ramonjd! This is testing well for me in the editor and on the front-end! 🎉  Since it looks like the `blockGap` isn't currently functional in Global Styles, would it be worth us removing / hiding the control there for the time being? (The fact that it doesn't currently do anything isn't because of this PR, of course)

I like the approach to concatenating the row and column values into the `gap` value for those blocks where we opt-in to the axial controls 👍. It's otherwise working really nicely for me!

ramonjd commented:
Thanks for taking this for another spin @andrewserong 🙇 

> would it be worth us removing / hiding the control there for the time being? (The fact that it doesn't currently do anything isn't because of this PR, of course)

Just so I know we're talking about the same thing 😄 Do you mean for blocks?

<img width="290" alt="Screen Shot 2022-03-21 at 3 28 09 pm" src="https://user-images.githubusercontent.com/6458278/159205833-4369bd1d-894a-4061-ae38-e10e9fbf7ffb.png">

This isn't having any effect for me either, and I've never noticed it before. So I totally agree!  😆 

It's turned off for [ROOT_BLOCK_SUPPORTS](https://github.com/WordPress/gutenberg/blob/e2f00ff5ed1f4042386a48eedf4eafef36c9eb29/packages/edit-site/src/components/global-styles/hooks.js#L156) as well, so I don't see any harm in removing it from the dimensions panel.

I'll throw up a PR.

andrewserong reviewed (approved):
LGTM, thanks for opening up the follow-up PR to remove the controls from the dimensions panel in global styles!

I suppose to make it neater before landing, we could remove the changes to [packages/edit-site/src/components/global-styles/dimensions-panel.js](https://github.com/WordPress/gutenberg/pull/37736/files#diff-27d286e3b583bb140e30590a8992de3d1282b24ee7436c24178f9d159b49e19c)?
